src/referral.cairo

```diff
@@ -0,0 +1,202 @@
+#[abi]
+trait IReferral {
+    #[view]
+    fn get_balance(sponsor_addr: starknet::ContractAddress) -> u256;
+    #[view]
+    fn owner() -> starknet::ContractAddress;
+    #[external]
+    fn claim(amount: u256);
+    #[external]
+    fn add_commission(amount: u256, sponsor_addr: starknet::ContractAddress);
+    #[external]
+    fn withdraw(addr: starknet::ContractAddress, amount: u256);
+    #[external]
+    fn upgrade(impl_hash: starknet::class_hash::ClassHash);
+    #[external]
+    fn upgrade_and_call(new_hash: starknet::class_hash::ClassHash, selector: felt252, calldata: Array<felt252>);
+}
+
+#[contract]
+mod Referral {
+    use starknet::ContractAddress;
+    use starknet::class_hash::ClassHash;
+    use starknet::{get_caller_address, get_contract_address, get_block_timestamp};
+
+    use debug::PrintTrait;
+
+    use referral::access::ownable::Ownable;
+    use referral::upgrades::upgradeable::Upgradeable;
+
+    // dispatchers
+    use referral::token::erc20::{ IERC20Dispatcher, IERC20DispatcherTrait };
+
+    struct Storage {
+        sponsor_balance: LegacyMap::<ContractAddress, u256>,
+        sponsor_comm: LegacyMap::<ContractAddress, u256>,
+        default_comm: u256,
+        min_claim: u256,
+        naming_contract: ContractAddress,
+        eth_contract: ContractAddress,
+    }
+
+    //
+    // Events
+    //
+
+    #[event]
+    fn on_claim(timestamp: u64, amount: u256, sponsor_addr: ContractAddress, ) {}
+
+    #[event]
+    fn on_commission(timestamp: u64, amount: u256, sponsor_addr: ContractAddress, caller: ContractAddress, ) {}
+
+
+    //
+    // Constructor
+    //
+
+    #[constructor]
+    fn constructor(
+        admin: ContractAddress, 
+        naming_addr: ContractAddress,
+        eth_addr: ContractAddress,
+        min_claim_amount: u256, 
+        share: u256
+    ) {
+        initializer(:admin);
+        naming_contract::write(naming_addr);
+        eth_contract::write(eth_addr);
+        min_claim::write(min_claim_amount);
+        default_comm::write(share);
+    }
+    
+    //
+    // View
+    //
+
+    #[view]
+    fn get_balance(sponsor_addr: ContractAddress) -> u256 {
+        sponsor_balance::read(sponsor_addr)
+    }
+
+    //
+    // External
+    //
+
+    #[external]
+    fn claim(amount: u256) {
+        assert(amount >= min_claim::read(), 'Amount is too low');
+        let sponsor_addr = get_caller_address();
+        let balance = sponsor_balance::read(sponsor_addr);
+        assert(balance >= amount, 'Amount greater than balance');
+
+        let contract_addr = get_contract_address();
+        let ERC20 = IERC20Dispatcher { contract_address: eth_contract::read() };
+        let contract_balance = ERC20.balance_of(contract_addr);
+        assert(contract_balance >= amount, 'Contract balance too low');
```

Comment:
I don't think we need this one, ERC20 contract will do it for us.  Having a custom message is a good idea though. Is it possible to surround ERC20.transferFrom with a with_attr equivalent, similar to what we could do in Cairo 0?

Reply:
I couldn't make it work with a custom error msg for now so I removed that check in the `claim` & `withdraw` functions.